core/src/main/java/io/grpc/internal/OutboundState.java

```diff
@@ -0,0 +1,294 @@
+   */
+  interface Handler {
+    /**
+     * Sends the given headers to the remote endpoint.
+     */
+    void sendHeaders(Metadata headers);
+
+    /**
+     * Sends the given message to the remote endpoint.
+     *
+     * @param frame the message to be sent.
+     * @param endOfStream if {@code true}, indicates that the frame should be marked as the end of stream.
+     * @param flush if {@code true}, indicates that any buffered IO should be flushed to the remote endpoint.
+     */
+    void sendMessage(@Nullable WritableBuffer frame, boolean endOfStream, boolean flush);
+
+    /**
+     * Sends the specified trailers to the remote endpoint. This is only called for server-side
+     * handlers.
+     *
+     * @param trailers the trailers to be sent
+     * @param headersSent if {@code true}, indicates that headers were previously sent.
+     */
+    void sendTrailers(Metadata trailers, boolean headersSent);
+
+    /**
+     * This indicates that the transport is now capable of sending additional messages
+     * without requiring excessive buffering internally. This event is
+     * just a suggestion and the application is free to ignore it, however doing so may
+     * result in excessive buffering within the transport.
+     */
+    void onReady();
+  }
+
+  private final boolean server;
+  private final Handler handler;
+  private final MessageFramer framer;
+  private Phase phase = Phase.HEADERS;
+  private boolean headersSent;
+  private Metadata stashedTrailers;
+
+  /**
+   * The number of bytes currently queued, waiting to be sent. When this falls below
+   * onReadyThreshold, {@link StreamListener#onReady()} will be called.
+   */
+  @GuardedBy("onReadyLock")
+  private int numSentBytesQueued;
+
+  private final Object onReadyLock = new Object();
+
+  OutboundState(WritableBufferAllocator bufferAllocator, Handler handler, boolean server) {
+    this.handler = checkNotNull(handler, "handler");
+    this.server = server;
+    MessageFramer.Sink frameHandler = new MessageFramer.Sink() {
+      @Override
+      public void deliverFrame(WritableBuffer frame, boolean endOfStream, boolean flush) {
+        onMessageFrame(frame, endOfStream, flush);
+      }
+    };
+    framer = new MessageFramer(frameHandler, bufferAllocator);
+  }
+
+  void setCompressor(Compressor c) {
+    // TODO(carl-mastrangelo): check that headers haven't already been sent.  I can't find where
+    // the client stream changes outbound phase correctly, so I am ignoring it.
+    framer.setCompressor(c);
+  }
+
+  /**
+   * Request the sending of headers. Must be called from the channel thread context.
+   */
+  void sendHeaders(Metadata headers) {
+    checkNotNull(headers, "headers");
+    phase(Phase.HEADERS);
+    headersSent = true;
+    handler.sendHeaders(headers);
+    phase(Phase.MESSAGE);
+  }
+
+  /**
+   * Request that a message be sent. Must be called from the channel thread context.
+   */
+  void sendMessage(InputStream message) {
```

Comment:
Yes, the application cannot call anything after close. But the application can't trigger anything after closing the framer because of state checking in ServerCallImpl.